### PR TITLE
KAFKA-4622: Consumer should handle group authorization errors in offset fetch

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:775]: I think previously we treated this as retriable because we know the metadata could simply be outdated and so you need to retry in case the topic finally appears. But looking at it from the perspective of auth, we just handle retriable errors by retrying forever. For auth stuff, that may never get resolved... And even in the case of topics that may be getting created, it could just be a fat-fingered config and the topic might never exist. And I think all you'll end up with in this case is debug-level logs that mention the error (if you're lucky enough to have them enabled).

This isn't really a problem with this patch, but something I realized when reviewing it. It might just be the same as https://issues.apache.org/jira/browse/KAFKA-3177 in which case maybe we can defer the solution to that bug, but it seems a likely source of seemingly mysterious issues for users that it might be easy to at least resolve an extra log message into a patch like this to help alleviate the issue.

**Review by ewencp — APPROVED** [review]
@hachikuji LGTM for what's mentioned in the JIRA, but see my comment. I'll leave the commit up to you based on what you think we should do about the issue I mentioned -- if you want to defer to that other JIRA, feel free to commit this PR as is.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:775]: @ewencp Thanks for the comments. Apologies for the long reply, but I've been thinking a bit about this as well after reviewing KIP-88. One interesting thing I found is that we actually only check topic existence when fetching offsets from zookeeper (offset fetch version 0). When fetching from Kafka, we return INVALID_OFFSET with no error code set if the user has Describe access to a topic which does not exist. So currently, the only way to get UNKNOWN_TOPIC_OR_PARTITION is if the user doesn't have Describe access to the topic. That probably explains why the current behavior has not caused any issues (that I'm aware of). 

Another related issue: when a topic is deleted, we do not remove corresponding offsets from the caches. Given the behavior above, that means the offsets remain fetchable. Generally the impact of this is minor (the offsets will eventually expire), but it could be bad if the topic was re-created quickly. In the normal case, if a consumer assigned itself a non-existing partition which had a still fetchable offset, then we'd get stuck in a loop waiting for the consumer to fetch the topic metadata. If we fixed the behavior above, then we should probably let the consumer continue trying to fetch offsets as long as it receives UNKNOWN_TOPIC_OR_PARTITION errors. Another (small) problem if we don't also fix the deletion problem is that the group cannot be cleaned up until all offsets have expired, yet the user would have no way to tell which offsets remain because the offset fetch would also return the unknown topic error. (By the way, if I recall correctly, there was a patch to fix the deletion problem, but the contributor wasn't too responsive. Probably worth trying to unblock that work.)

As to the general problem of the client not being able to distinguish between non-existing topics and authorization failures.... I agree that it sucks, but I don't see much we can do about it given the way the clients currently handles unknown topics. My hope is that we can deprecate the "retry forever until the topic exists" behavior over time as auto-creation of topics itself is deprecated. In the short term, maybe bumping up the logging is all we can do. I'm actually guessing it's uncommon in practice to not have Describe access to topics, but who knows?

Anyway, I think there are enough related issues, that we should probably deal with this in a separate JIRA, so I'm inclined to merge this patch.

**Review by ijuma — APPROVED** [review]
Thanks for the PR, LGTM.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:775]: I agree that this is something that we should fix, but it's a broader issue and affects many request types, not just this one.

**Comment by ijuma** [comment]
Since Ewen said he was OK to merge this as is based on Jason's input, I went ahead and did it.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:775]: FYI: I created https://issues.apache.org/jira/browse/KAFKA-4665 for the inconsistencies mentioned above.

**Pull request closed without merging**
